internal/dinosaur/pkg/gitops/config.go

```diff
@@ -54,6 +104,134 @@ func ValidateConfig(config Config) field.ErrorList {
+	return errs
+}
+
+func validateAdditionalAuthProviders(path *field.Path, providers []AuthProviderAddition) field.ErrorList {
+	var errs field.ErrorList
+	for i, additionalProvider := range providers {
+		errs = append(errs, validateAdditionalAuthProvider(path.Index(i), additionalProvider)...)
+	}
+	return errs
+}
+
+func validateAdditionalAuthProvider(path *field.Path, provider AuthProviderAddition) field.ErrorList {
+	var errs field.ErrorList
+
+	if provider.InstanceID == "" {
+		errs = append(errs, field.Required(path.Child("instanceId"), "instance ID is required"))
+	}
+	authProviderPath := path.Child("authProvider")
+	errs = append(errs, validateAuthProvider(authProviderPath, provider.AuthProvider)...)
+
+	return errs
+}
+
+func validateAuthProvider(path *field.Path, provider *AuthProvider) field.ErrorList {
+	var errs field.ErrorList
+	if provider == nil {
+		errs = append(errs, field.Required(path, "auth provider spec is required"))
+		return errs
+	}
+	errs = append(errs, validateAuthProviderName(path.Child("name"), provider.Name)...)
+	errs = append(errs, validateAuthProviderGroups(path.Child("groups"), provider.Groups)...)
+	errs = append(errs, validateAuthProviderRequiredAttributes(path.Child("requiredAttributes"), provider.RequiredAttributes)...)
+	errs = append(errs, validateAuthProviderClaimMappings(path.Child("claimMappings"), provider.ClaimMappings)...)
+	// Empty config means that the config will be copied over from the default auth provider.
+	if provider.OIDC != nil {
+		errs = append(errs, validateAuthProviderOIDCConfig(path.Child("oidc"), provider.OIDC)...)
+	}
+	return errs
+}
+
+func validateAuthProviderClaimMappings(path *field.Path, claimMappings []AuthProviderClaimMapping) field.ErrorList {
+	var errs field.ErrorList
+	for i, claimMapping := range claimMappings {
+		errs = append(errs, validateAuthProviderClaimMapping(path.Index(i), claimMapping)...)
+	}
+	return errs
+}
+
+func validateAuthProviderRequiredAttributes(path *field.Path, requiredAttributes []AuthProviderRequiredAttribute) field.ErrorList {
+	var errs field.ErrorList
+	for i, requiredAttribute := range requiredAttributes {
+		errs = append(errs, validateAuthProviderRequiredAttribute(path.Index(i), requiredAttribute)...)
+	}
+	return errs
+}
+
+func validateAuthProviderGroups(path *field.Path, groups []AuthProviderGroup) field.ErrorList {
+	var errs field.ErrorList
+	for i, group := range groups {
+		errs = append(errs, validateAuthProviderGroup(path.Index(i), group)...)
+		for j, anotherGroup := range groups {
+			if i != j {
+				if group.Role == anotherGroup.Role && group.Key == anotherGroup.Key && group.Value == anotherGroup.Value {
+					groupDescription := fmt.Sprintf("%s-%s-%s", group.Role, group.Key, group.Value)
+					errs = append(errs, field.Duplicate(path.Index(i), groupDescription))
+				}
+			}
+		}
+	}
+	return errs
+}
```

Comment:
instead of a N^2 approach to finding duplicates, you can maintain a list of "seen groups"

```suggestion
func validateAuthProviderGroups(path *field.Path, groups []AuthProviderGroup) field.ErrorList {
	var errs field.ErrorList
	var seenGroups = make(map[AuthProviderGroup]struct{}, 0, len(groups))
	for i, group := range groups {
	    groupPath := path.Index(i)
	    if _, ok := seenGroups[group]; ok {
	        errs = append(errs, field.Duplicate(groupPath, fmt.Sprintf("duplicate group %v", group)))
	        continue
	     }
	     seenGroups[group] = {}
	     errs = append(errs, validateAuthProviderGroup(groupPath, group)...)
	}
	return errs
}
```